### PR TITLE
Added support for processor affinity for afl-fuzz workers

### PR DESCRIPTION
**Pull request by yoava333**
this patch adds support for affinity & fixes https://github.com/ivanfratric/winafl/issues/91

### CODE REVIEW

File: afl-fuzz.c [hunk 1]
```diff
@@ -383,6 +385,151 @@ static void shuffle_ptrs(void** ptrs, u32 cnt) {
...
+      return i;
+    }
+  }
+
+  return 0;
+}
+
+static void bind_to_free_cpu(void) {
+
+  u8 cpu_used[4096];
```
Comment (ivanfratric) [inline afl-fuzz.c]: cpu_used[64]?

File: afl-fuzz.c [hunk 2]
```diff
@@ -383,6 +385,151 @@ static void shuffle_ptrs(void** ptrs, u32 cnt) {
 
 }
 
+static void usleep(u64 usec)
```
Comment (ivanfratric) [inline afl-fuzz.c]: Is this really needed? Why won't the plain old Sleep() work, considering the intervals involved are not sufficiently small to justify needing microsecond precision.

**Comment by yoava333** [comment]
Addressed your comments for the code.

I ran two benchmarks:
* on a CPU E5-2630 v4 @ 2.20GHz (20 core, HT) - I saw ~10% increase - from ~450/s to ~500/s
* on a my laptop i7-7700HQ (8 cores, HT):
  - 2 fuzzers without affinity: ~1000/s
  - 2 fuzzers with affinity to Hyper threaded cores: ~950/s, ~650/s
  - 2 fuzzers with affinity to different hyper threaded cores: ~1180/s (~18% increase)
  - 2 fuzzers on the same core ~480/s

So it depends if you are using all your cores, if not this might degrade performance since sequential cores are usually hyper threaded cores.

**Comment by ivanfratric** [comment]
That's an interesting observation about sequential cores. As a workaround, do you think it makes sense to first try to pin to even cores (0, 2, 4, ...) and if none are available do a pass over odd cores?

**Comment by yoava333** [comment]
I think it does make sense, it doesn't hurt and in my tests (on both machines) it showed better performance when using subset of my cores. I've updated the code.

**Comment by ivanfratric** [comment]
Thanks!

**Comment by 0vercl0k** [comment]
This is super cool btw - never expected to have an 18% gap between both
configurations :). Thanks @ivanfratric and @yoava333!

Cheers

2018-03-19 5:00 GMT-07:00 ivanfratric <notifications@github.com>:

**Pull request merged**